RecoHI/HiJetAlgos/plugins/HiPuRhoProducer.cc

```diff
@@ -0,0 +1,645 @@
+
+  bool postOrphan_;
+
+  bool dropZeroTowers_;
+  int medianWindowWidth_;
+  double minimumTowersFraction_;
+  double nSigmaPU_;  // number of sigma for pileup
+  double puPtMin_;
+  double radiusPU_;    // pileup radius
+  double rParam_;      // the R parameter to use
+  edm::InputTag src_;  // input constituent source
+  double towSigmaCut_;
+
+  std::vector<edm::Ptr<reco::Candidate>> inputs_;  // input candidates
+  ClusterSequencePtr fjClusterSeq_;                // fastjet cluster sequence
+  JetDefPtr fjJetDefinition_;                      // fastjet jet definition
+
+  std::vector<fastjet::PseudoJet> fjInputs_;          // fastjet inputs
+  std::vector<fastjet::PseudoJet> fjJets_;            // fastjet jets
+  std::vector<fastjet::PseudoJet> fjOriginalInputs_;  // to back-up unsubtracted fastjet inputs
+
+  CaloGeometry const* geo_ = nullptr;  // geometry
+  std::vector<HcalDetId> allgeomid_;   // all det ids in the geometry
+
+  int ietamax_;                               // maximum eta in geometry
+  int ietamin_;                               // minimum eta in geometry
+  std::map<int, int> ntowersWithJets_;        // number of towers with jets
+  std::map<int, int> geomtowers_;             // map of geometry towers to det id
+  std::map<int, double> esigma_;              // energy sigma
+  std::map<int, double> emean_;               // energy mean
+  std::map<int, std::vector<double>> eTop4_;  // energy mean
+
+  edm::EDGetTokenT<reco::CandidateView> input_candidateview_token_;
+};
+
+HiPuRhoProducer::HiPuRhoProducer(const edm::ParameterSet& iConfig)
+    : dropZeroTowers_(iConfig.getParameter<bool>("dropZeroTowers")),
+      medianWindowWidth_(iConfig.getParameter<int>("medianWindowWidth")),
+      minimumTowersFraction_(iConfig.getParameter<double>("minimumTowersFraction")),
+      nSigmaPU_(iConfig.getParameter<double>("nSigmaPU")),
+      puPtMin_(iConfig.getParameter<double>("puPtMin")),
+      radiusPU_(iConfig.getParameter<double>("radiusPU")),
+      rParam_(iConfig.getParameter<double>("rParam")),
+      src_(iConfig.getParameter<edm::InputTag>("src")),
+      towSigmaCut_(iConfig.getParameter<double>("towSigmaCut")) {
+  input_candidateview_token_ = consumes<reco::CandidateView>(src_);
+
+  //register your products
+  produces<std::vector<double>>("mapEtaEdges");
+  produces<std::vector<double>>("mapToRho");
+  produces<std::vector<double>>("mapToRhoMedian");
+  produces<std::vector<double>>("mapToRhoExtra");
+  produces<std::vector<double>>("mapToRhoM");
+  produces<std::vector<int>>("mapToNTow");
+  produces<std::vector<double>>("mapToTowExcludePt");
+  produces<std::vector<double>>("mapToTowExcludePhi");
+  produces<std::vector<double>>("mapToTowExcludeEta");
+}
+
+// ------------ method called to produce the data  ------------
+void HiPuRhoProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  setupGeometryMap(iEvent, iSetup);
+
+  inputs_.clear();
+  fjInputs_.clear();
+  fjJets_.clear();
```

Comment:
How bit can these containers become? In general we should not keep unnecessary stuff in memory from one event to the next.

Reply:
For some reason quite a few jet producers clears this kind of containers at the beginning of an event. I don't see why they can't be cleared at the end instead.

Reply:
Thanks. Maybe this hints for an opportunity for a later cleanup in all affected jet producers (@slava77 @perrotta @jpata). In principle there should be no reason to store such data in member variables in the first place (especially given that `.clear()` does not free the memory).